uefi/src/mem.rs

```diff
@@ -0,0 +1,136 @@
+//! This is a utility module with helper methods for allocations/memory.
+
+use crate::ResultExt;
+use crate::{Result, Status};
+use ::alloc::{alloc, boxed::Box};
+use core::alloc::Layout;
+use core::fmt::Debug;
+use core::ptr::NonNull;
+use core::slice;
+use uefi::data_types::Align;
+use uefi::Error;
+
+/// Helper to return owned versions of certain UEFI data structures on the heap, hence, in a
+/// [`Box`]. It is intended to wrap low-level UEFI functions off this crate that
+/// - consume a mutable reference to a buffer that needs to be filled with some data of type,
+/// - return a mutable typed reference (which points to the same memory as the input buffer) on
+///   success, and
+/// - return the required buffer size on error.
+pub fn make_boxed<
+    'a,
+    Data: Align + ?Sized + Debug + 'a,
+    F: FnMut(&'a mut [u8]) -> Result<&'a mut Data, Option<usize>>,
+>(
+    mut fetch_data_fn: F,
+) -> Result<Box<Data>> {
+    let required_size = match fetch_data_fn(&mut []).map_err(Error::split) {
+        // This is the expected case: the empty buffer passed in is too
+        // small, so we get the required size.
+        Err((Status::BUFFER_TOO_SMALL, Some(required_size))) => Ok(required_size),
+        // Propagate any other error.
+        Err((status, _)) => Err(Error::from(status)),
+        // Success is unexpected, return an error.
+        Ok(_) => Err(Error::from(Status::UNSUPPORTED)),
+    }?;
+
+    // We add trailing padding because the size of a rust structure must
+    // always be a multiple of alignment.
+    let layout = Layout::from_size_align(required_size, Data::alignment())
+        .unwrap()
+        .pad_to_align();
+
+    // Allocate the buffer.
+    let heap_buf: NonNull<u8> = unsafe {
+        let ptr = alloc::alloc(layout);
+        match NonNull::new(ptr) {
+            None => return Err(Status::OUT_OF_RESOURCES.into()),
+            Some(ptr) => ptr,
+        }
+    };
+
+    // Get the file info using the allocated buffer for storage.
+    let data: Result<&mut Data> = {
+        let buffer = unsafe { slice::from_raw_parts_mut(heap_buf.as_ptr(), layout.size()) };
```

Comment:
Could we change `layout.size()` to `required_size`? I'm not quite clear on whether the layout's size could be different due to the call to `pad_to_align`.

Reply:
good idea